android/src/main/java/com/twiliorn/library/PatchedVideoView.java

```diff
@@ -45,6 +64,108 @@ public void run() {
+                nv21Data[ySize + y * chromaStride + 2 * x + 1] = uValue;
+            }
+        }
+        return nv21Data;
+    }
+
+    private void sendEvent(String event, WritableMap params) {
+        ((ReactContext) this.getContext()).getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
+                .emit(event, params);
+    }
+
+    private void saveVideoFrame(VideoFrame frame) {
+        frame.retain();
+        Context context = this.getContext();
+        String fileId = UUID.randomUUID().toString();
+        String filename = "rntframe-" + fileId + ".jpeg";
+        VideoFrame.I420Buffer i420Buffer = frame.getBuffer().toI420();
+        final int width = i420Buffer.getWidth();
+        final int height = i420Buffer.getHeight();
+        byte[] nv21Data = createNV21Data(i420Buffer);
+        YuvImage yuvImage = new YuvImage(nv21Data, ImageFormat.NV21, width, height, null);
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        yuvImage.compressToJpeg(new Rect(0, 0, width, height), 100, out);
+        byte[] rawImageBytes = out.toByteArray();
+        byte[] imageBytes = rotateJPEGByteArray(rawImageBytes, 90f);
+        try (FileOutputStream fos = context.openFileOutput(filename, Context.MODE_PRIVATE)) {
+            // write image to disk
+            fos.write(imageBytes);
+            Log.d(TwilioPackage.TAG, "saved frame to " + filename);
+            // send event to JS w/ filename
+            WritableMap params = Arguments.createMap();
+            params.putString("filename", filename);
+            sendEvent("onFrameCaptured", params);
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException(e);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        } finally {
+            frame.release();
+        }
+    }
+
+    private byte[] rotateJPEGByteArray(
+            byte[] data,
+            float rotation
+    ) {
+        if (rotation == 0f) {
+            return data;
+        }
+         else {
+            Bitmap sourceBitmap = BitmapFactory.decodeByteArray(data, 0, data.length);
+            Bitmap rotatedBitmap = rotateBitmap(sourceBitmap, rotation);
+            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+            rotatedBitmap.compress(Bitmap.CompressFormat.JPEG, 100, byteArrayOutputStream);
+            sourceBitmap.recycle();
+            rotatedBitmap.recycle();
+            return byteArrayOutputStream.toByteArray();
+        }
+    }
+
+    private Bitmap rotateBitmap(
+            Bitmap sourceBitmap,
+            float rotation
+    ) {
+        Matrix matrix = new Matrix();
+        matrix.postRotate(rotation);
+        return Bitmap.createBitmap(sourceBitmap, 0, 0, sourceBitmap.getWidth(), sourceBitmap.getHeight(), matrix, true);
+    }
+
+    public void notifyCaptureFrame() {
+        this.captureThisFrame = true;
```

Comment:
❓ Do you know if this `notifyCaptureFrame()` method is called on the same thread as `onFrame()`? If not, then you will need to make `this.captureThisFrame` thread-safe using `AtomicBoolean`.

Reply:
I do not know. Should I make it thread safe to be safe?

Reply:
I would use `AtomicBoolean` to be safe. I didn't fully trace the code, but it seemed likely to me when I last looked at this that the threads could be different. `AtomicBoolean` is a low-effort solution when you're dealing with just a single boolean variable. If you get to the point where you have multiple variables that need to be kept in sync, or a single variable with a non-primitive type, then it's worth figuring out what's actually going on and the best way to keep things thread-safe.

Reply:
Ok cool. I noticed some of the twilio examples I've been getting some inspiration from have been doing the same thing

https://github.com/twilio/video-quickstart-android/blob/master/exampleAdvancedCameraCapturer/src/main/java/com/twilio/video/examples/advancedcameracapturer/Photographer.kt#L14